### PR TITLE
Implement optimization for getting diagnostics for a SyntaxTree with Global Using directives

### PR DESCRIPTION
**Pull request by AlekseyTs**
Test plan: https://github.com/dotnet/roslyn/issues/51307

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review.

File: src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:161]: >UsageOfUsingsRecorededInTrees [](start = 47, length = 29)

Nit: typo, `UsageOfUsingsRecordedInTrees`

File: src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:2902]: >ed [](start = 25, length = 2)

Nit: should this be `register` instead of `registered`?

File: src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:2922]: Does this need to be a `Volatile.Write` to ensure that it can't be reordered on another thread that is doing the interlocked check below?

**Comment by 333fred** [comment]
Done review pass (commit 2). I'd like to make sure the threading question is answered before signing off.

File: src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:2902]: > should this be register instead of registered?

I really struggled with this name. We are not registering usage here, we are registering the fact that we registered usage. Dodn't want to use "register" twice

---
In reply to: [601865670](https://github.com/dotnet/roslyn/pull/52132#discussion_r601865670) [](ancestors = 601865670)

File: src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:2922]: > Does this need to be a Volatile.Write to ensure that it can't be reordered on another thread that is doing the interlocked check below?

I am not sure what is the concern here. Interlocked is guaranteed to operate on the most recent value, I believe.

---
In reply to: [601866669](https://github.com/dotnet/roslyn/pull/52132#discussion_r601866669) [](ancestors = 601866669)

Comment (333fred) [reply src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:2922]: Alright. I think I've convinced myself that there's not a problem here.

---
In reply to: [601872182](https://github.com/dotnet/roslyn/pull/52132#discussion_r601872182) [](ancestors = 601872182,601866669)

**Review by 333fred — APPROVED** [review]
LGTM (commit 2), modulo the typo (field and prop)

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review, need a second sign-off.

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review, need a second sign-off.

File: src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs
Comment (cston) [inline src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:2851]: >recordUsageOfUsingsInAllTrees [](start = 20, length = 29)

Does this mean we're compiling all trees in the compilation twice if there are any global usings in the compilation?

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:2851]: > Does this mean we're compiling all trees in the compilation twice if there are any global usings in the compilation?

This function is not called during normal build process. It is only called when a client requests diagnostics for a specific syntax tree through SemanticModel. That tree wiil always be compiled. When it contains global usings, we will also compile other trees, unless we know that we already did that for them.  

---
In reply to: [602637179](https://github.com/dotnet/roslyn/pull/52132#discussion_r602637179) [](ancestors = 602637179)

**Review by cston — APPROVED** [review]

**Pull request merged**
